gusto/advection.py

```diff
@@ -74,46 +68,91 @@ def __init__(self, state, field, equation=None, *, solver_parameters=None,
-                if self.limiter is not None:
-                    self.x_brok_interpolator = Interpolator(self.xdg_out, x_brok)
-                    self.x_out_projector = Recoverer(x_brok, self.x_projected)
-        else:
-            self.embedded_dg = False
-            fs = field.function_space()
 
-        # setup required functions
-        self.fs = fs
-        self.dq = Function(fs)
-        self.q1 = Function(fs)
+        if options.name is "recovered":
+            V_rec = options.recovered_space
+            V_brok = options.broken_space
+
+            # set up the necessary functions
+            self.x_in = Function(field.function_space())
+            x_adv = Function(self.fs)
+            x_rec = Function(V_rec)
+            x_brok = Function(V_brok)
+
+            # set up interpolators and projectors
+            self.x_adv_interpolator = Interpolator(self.x_in, x_adv)  # interpolate before recovery
+            self.x_rec_projector = Recoverer(x_adv, x_rec)  # recovered function
+            # when the "average" method comes into firedrake master, this will be
+            # self.x_rec_projector = Projector(self.x_in, equation.Vrec, method="average")
+            self.x_brok_projector = Projector(x_rec, x_brok)  # function projected back
+            self.xdg_interpolator = Interpolator(self.x_in + x_rec - x_brok, self.xdg_in)
+            if self.limiter is not None:
+                self.x_brok_interpolator = Interpolator(self.xdg_out, x_brok)
+                self.x_out_projector = Recoverer(x_brok, self.x_projected)
+                # when the "average" method comes into firedrake master, this will be
+                # self.x_out_projector = Projector(x_brok, self.x_projected, method="average")
+
+    def pre_apply(self, x_in, discretisation_option):
+        """
+        Extra steps to the apply method for the recovered advection scheme.
+        This provides an advection scheme for the lowest-degree family
+        of spaces, but which has second order numerical accuracy.
+
+        :arg x_in: the input set of prognostic fields.
+        """
+        if discretisation_option == "embedded_dg":
+            try:
+                self.xdg_in.interpolate(x_in)
+            except NotImplementedError:
+                self.xdg_in.project(x_in)
+
+        elif discretisation_option == "recovered":
+            self.x_in.assign(x_in)
+            self.x_adv_interpolator.interpolate()
+            self.x_rec_projector.project()
+            self.x_brok_projector.project()
+            self.xdg_interpolator.interpolate()
+
+    def post_apply(self, x_out, discretisation_option):
+        """
+        The projection steps for the recovered advection scheme,
+        used for the lowest-degree sets of spaces. This returns the
+        field to its original space, from the space the embedded DG
+        advection happens in. This step acts as a limiter.
```

Comment:
I realise that I wrote this, but technically this step is only limited if `self.limiter is not None`